tests/gcoap_fileserver/tests/01-run.py

```diff
@@ -0,0 +1,121 @@
+#!/usr/bin/env python3
+
+# Copyright (C) 2022 Benjamin Valentin
+#
+# This file is subject to the terms and conditions of the GNU Lesser
+# General Public License v2.1. See the file LICENSE in the top level
+# directory for more details.
+
+import subprocess
+import time
+
+from subprocess import Popen
+from riotctrl_shell.gnrc import GNRCICMPv6Echo, GNRCICMPv6EchoParser
+from riotctrl_shell.netif import Ifconfig
+from riotctrl.ctrl import RIOTCtrlBoardFactory
+from riotctrl_ctrl import native
+from riotctrl_shell.netif import IfconfigListParser
+
+PARSERS = {
+    "ping6": GNRCICMPv6EchoParser(),
+    "ifconfig": IfconfigListParser(),
+}
+
+
+class RIOTCtrlAppFactory(RIOTCtrlBoardFactory):
+
+    def __init__(self):
+        super().__init__(board_cls={
+            'native': native.NativeRIOTCtrl,
+        })
+        self.ctrl_list = list()
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *exc):
+        for ctrl in self.ctrl_list:
+            ctrl.stop_term()
+
+    def get_shell(self, application_directory='.', env={'BOARD': 'native'}):
+        # retrieve a RIOTCtrl Object
+        ctrl = super().get_ctrl(
+            env=env,
+            application_directory=application_directory
+        )
+        # append ctrl to list
+        self.ctrl_list.append(ctrl)
+        # start terminal
+        ctrl.start_term()
+        # return ctrl with started terminal
+        return Shell(ctrl)
+
+    def get_shells(self, num=1):
+        terms = []
+        for i in range(num):
+            terms.append(self.get_shell())
+        return terms
+
+
+class Shell(Ifconfig, GNRCICMPv6Echo):
+    pass
+
+
+def first_netif_and_addr_by_scope(ifconfig_out, scope):
+    netifs = PARSERS["ifconfig"].parse(ifconfig_out)
+    key = next(iter(netifs))
+    netif = netifs[key]
+    return (
+        key,
+        [addr["addr"] for addr in netif["ipv6_addrs"] if addr["scope"] == scope][0],
+    )
+
+
+def global_addr(ifconfig_out):
+    return first_netif_and_addr_by_scope(ifconfig_out, "global")
+
+
+def test_linear_topology(factory, zep_dispatch):
+
+    # linear topology with 2 nodes and 30% packet loss
+    topology = ("A B 0.7\n")
```

Comment:
Some back-of-the-envelope math: We're sending a single request in 600 byte / 16 byte per block = 38 flights. Each flight has a 0.7**2 ~= 0.5 chance of failing, and 3 retransmissions (4 total, give or take) give every flight a 0.85 success rate. It's rather certain that of these 38 flights, one would fail and thus the transfer fails (P(all succeed) ~= 0.001).

I trust you tested this, so why did it succeed, or is my math wrong?

Reply:
Chance if failure for a single block is 2*0.3⁴ with 4 retransmissions, so 1.62%.
So the chance of success with 38 blocks is 0.9838^38 = 53.76% - which is indeed a bit on the low side for a test that should run on CI.

Reply:
Yikes -- I did the calculation for the success rate wrong. 0.5 ** (1/4) rather than 1 - (1 - 0.5)**4. How embarrassing.

Still, yes, the total 50% are relying on the current rerun regime. Setting the block size to 128 might be the best tuning knob; brings chances from 53% to 92%. Still, the README (and any semantic metadata about tests, which we don't have) should indicate that this test expects this probability of failure.

Reply:
Let's spicy up the test by adding an intermediate hop, but bump up the retransmissions to 10 so we get 4*0.25^10 chance of failing a single block, which is low enough.